### PR TITLE
build(regression-tests): refactor to support weekly LTS CRON

### PR DESCRIPTION
**Pull request by gunnar-solo**
# Description
In principal, there are only 2 _major_ changes done by this PR.
1) a refactor of the `regression-tests.yaml` workflow, in which most of its logic is extracted into a [local action](https://github.com/orgs/community/discussions/26245#discussioncomment-3250935)
2) heavy editing of the `nightly-tests.yaml` workflow to become a cron orchestrator

# Design Choices (and consequences)
The choices I made for designing this Github Action scheduler came from a combination of...
* [There isn't a nice way to reference the outputs of all jobs in a github actions matrix](https://github.com/community/community/discussions/17245#discussioncomment-3814009)
* [Github Actions cron scheduling only works on the default branch](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule)

So!  Easy(ish) enough.  We have a scheduler to schedule differently-defined jobs within the same workflow as appropriate:
```yaml
    - cron: "0 5 * * 1-5" # weekdays @ 00:00 EST, run tests against dev
    - cron: "6 6 * * 1"   # monday   @ 01:06 EST, run expanded tests against v1.13.x 
    - cron: "7 7 * * 1"   # monday   @ 02:07 EST, run expanded tests against v1.12.x
    - cron: "8 8 * * 1"   # monday   @ 03:08 EST, run expanded tests against v1.11.x
    - cron: "9 9 * * 1"   # monday   @ 04:09 EST, run expanded tests against v1.10.x
...
jobs:
  regression_tests_dev:
      if: github.event_name == 'workflow_dispatch' || github.event.schedule == '0 5 * * 1-5'
...
  regression_tests_13:
    if: github.event.schedule == '6 6 * * 1'
...
```

Each job in said scheduler consumes the **newly-created regression-tests action**, reducing code copy/paste.  There is (at least) one unfortunate consequence of this design pattern, however:

⚠️⚠️Each LTS branch must undergo a similar refactor to `regression-tests.yaml` in order for this to be functional⚠️⚠️

This is because of the CI drift between branches, as well as the inherent logic problem of a [local action](https://github.com/orgs/community/discussions/26245#discussioncomment-3250935) existing on one branch (unless explicitly backported).  Follow-up commits will do similar refactor changes that this PR has done to LTS branches `v1.10.x`, `v1.11.x`, `v1.12.x`, and `v1.13.x`.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: .github/workflows/nightly-tests.yaml [hunk 1]
```diff
@@ -1,131 +1,128 @@
...
-        SKIP_DOCKER: 'true'
-      run: |
-        ./ci/deploy-to-kind-cluster.sh
-    - name: Testing - kube e2e regression tests
-      env:
-        KUBE2E_TESTS: ${{ matrix.kube-e2e-test-type }}
-        GITHUB_TOKEN: ${{ github.token }}
-        ACK_GINKGO_RC: true
-        ACK_GINKGO_DEPRECATIONS: 1.16.5
-        RELEASED_VERSION: "LATEST"
```
Comment (elcasteel) [inline .github/workflows/nightly-tests.yaml:90]: Are we setting this anywhere?

File: .github/actions/regression-tests/action.yaml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    with:
+      path: ~/go/pkg/mod
+      key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+      restore-keys: |
+        ${{ runner.os }}-go-
+  - uses: engineerd/setup-kind@v0.5.0
+    with:
+      # We rely on the `deploy-to-kind-cluster` script to create a kind cluster
+      skipClusterCreation: true
+      version: v0.17.0
```
Comment (sam-heilbron) [inline .github/actions/regression-tests/action.yaml:44]: I believe you exposed this as an option in the matrix, should this be `${{ matrix.kube-version.kubectl }}`

File: .github/workflows/regression-tests.yaml [hunk 3]
```diff
@@ -7,10 +7,6 @@ on:
 
 env:
   VERSION: '1.0.0-ci'
-  HELM_VERSION: 'v3.6.0'
-  KUBECTL_VERSION: 'v1.24.7'
-  CLUSTER_NODE_VERSION: 'v1.24.7@sha256:5c015142d9b60a0f6c45573f809957076514e38ec973565e2b2fe828b91597f5'
-  GITHUB_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
```
Comment (sam-heilbron) [inline .github/workflows/regression-tests.yaml:13]: I think that upgrade tests actually require this token to be set. @ianmacclancy may have more context about why

i notice that on a recent run: https://github.com/solo-io/gloo/actions/runs/3913732719/jobs/6691072013 we fail in the same way that I think occurs without this auth set.

File: .github/workflows/regression-tests.yaml [hunk 4]
```diff
@@ -55,119 +49,47 @@ jobs:
 
         pr_comment_skip=${{ steps.skip-ci-comment.outputs.value }}
         if [[ ! -z $pr_comment_skip && $pr_comment_skip = true ]] ; then
-          should_pass=true
+          should_run=false
```
Comment (sam-heilbron) [inline .github/workflows/regression-tests.yaml:52]: Why this change? 

As an aside, some of this setup code/sig code doesn't do anything

Comment (gunnar-solo) [reply .github/workflows/regression-tests.yaml:52]: `should_pass=true` as a concept was a variable sprinkled at the start of every `step` in the previous regression tests.  As far as I could rationalize, `should_run` accomplished what we wanted on its own

this is actually the main bit I REALLY wanted your review for--for context on what it was accomplishing

Comment (sam-heilbron) [reply .github/workflows/regression-tests.yaml:53]: Yeah since /sig-ci doesn't do anything we can probably conslidate into 1 variable. The original intention was that we can have jobs which:
- Do not run certain steps
- Run certain steps but "auto-pass them"
- Run steps and let the steps determine whether it was a success/failure

I like your change!

**Review by sam-heilbron — APPROVED** [review]

**Review by sam-heilbron — APPROVED** [review]

**Review by elcasteel — APPROVED** [review]
